### ISSUE TITLE
Language model with 4 lstm layers is not learning 

### ISSUE THREAD

**Issue by ahmedalbahnasawy** [espnet/espnet#1285]
I'm training RNNLM my conf file is similar to the default one in librispeech repo but the loss looks weird 
![Screen Shot 2019-10-17 at 2 37 38 PM](https://user-images.githubusercontent.com/50653875/67001977-efa41700-f0eb-11e9-9cbb-8393f5b4081a.png)

**Comment by zqs01** [comment]
I also have the same question.
![图片](https://user-images.githubusercontent.com/40491290/67140539-22861080-f28e-11e9-8b08-f3b0c2845eff.png)
I use the default configuration:
`layer: 1         # 2 for character LMs

unit: 1000       # 650 for character LMs
opt: sgd          # adam for character LMs
sortagrad: 0 # Feed samples from shortest to longest ; -1: enabled for all epochs, 0: disabled, other: enabled for 'other' epochs
batchsize: 300    # 1024 for character LMs
epoch: 20        # number of epochs
patience: 3
maxlen: 40        # 150 for character LMs
`

**Comment by ahmedalbahnasawy** [comment]
my conf file is :
layer: 4 "lstm"
unit: 1024
opt: adam       # or adam
sortagrad: 0   # Feed samples from shortest to longest ; -1: enabled for all epochs, 0: disabled, other: enabled for 'other' epochs
batchsize: 256 # batch size in LM training
epoch: 20      # if the data size is large, we can reduce this
patience: 3
maxlen: 40     # if sentence length > lm_maxlen, lm_batchsize is automatically reduced
dropout-rate: 0.2
for sentencepiece unigram unit

**Comment by ahmedalbahnasawy** [comment]
@kan-bayashi could you check the conf file please

**Comment by kan-bayashi** [comment]
@ahmedalbahnasawy 
Sorry for the late reply. 
We are still fighting with ICASSP papers...
From your config, it seems that the biggest difference is an optimizer.
Did you try SGD instead of adam?

@zqs01 What is your task?

@ShigekiKarita and @sw005320 may give useful comments to you after the end of ICASSP...

**Comment by sw005320** [comment]
Compared with https://github.com/espnet/espnet/blob/master/egs/librispeech/asr1/conf/tuning/lm_large.yaml there are nother differences I found in addition to the optimizer, e.g., dropout, batchsize, etc. I recommend you to follow our default lm configs before you customize it by yourself.

**Comment by ahmedalbahnasawy** [comment]
okay thank you @sw005320 and @kan-bayashi I will start training by following the same conf file

**Comment by sw005320** [comment]
Thanks. Please let us know if you still have some issues.

**Comment by ahmedalbahnasawy** [comment]
<img width="828" alt="Screen Shot 2019-10-31 at 10 11 19 AM" src="https://user-images.githubusercontent.com/50653875/67923312-e96f6980-fbc6-11e9-9891-a4c9ee54efe8.png">
it looks better than the first time but still not healthy. i used the default conf file for libri. Furthermore, i 'm training librispeech corpus only.
could you till me how it should look like

**Comment by sw005320** [comment]
Can you show us both the training and validation scores (perplexity)? I usually don't check the training loss only as it does not tell us enough information. Also, I do not think this looks wrong. 

FYI, this is the behavior of the validation scores for 6 epochs that I have
```
        "validation/main/loss": 3.727457046508789,
        "val_perplexity": 41.573254777575066
        "validation/main/loss": 3.634854316711426,
        "val_perplexity": 37.89633163275677
        "validation/main/loss": 3.6004064083099365,
        "val_perplexity": 36.61311129311776
        "validation/main/loss": 3.5864903926849365,
        "val_perplexity": 36.107131441498424
        "validation/main/loss": 3.5738625526428223,
        "val_perplexity": 35.65404314540044
        "validation/main/loss": 3.5689377784729004,
        "val_perplexity": 35.478886691665714
```

**Comment by ahmedalbahnasawy** [comment]
Thank you again @sw005320 i will show you the training and validation scores once i finish training the ASR

**Comment by ahmedalbahnasawy** [comment]
Hi @sw005320 @kan-bayashi 
i'm still facing the same problem
<img width="938" alt="Screen Shot 2020-03-29 at 4 49 28 PM" src="https://user-images.githubusercontent.com/50653875/78139632-dee92280-7439-11ea-90fc-fa69ce9c0000.png">
<img width="851" alt="Screen Shot 2020-04-01 at 4 46 37 PM" src="https://user-images.githubusercontent.com/50653875/78140079-7ea6b080-743a-11ea-9485-e24ab36deb0b.png">
i used those two config file 
[lm.txt](https://github.com/espnet/espnet/files/4415299/lm.txt)
[lm2.txt](https://github.com/espnet/espnet/files/4415300/lm2.txt)

The only different from default lm configs is number of layers
I used the acoustic text to train LM and the size of the train file is 123M 
Do you think replacing RNNLM or Transformer with Ngram model could be better or i missing something

**Comment by sw005320** [comment]
This perplexity range looks normal. The loss is also gradually decreased. It is well trained I believe. If you have some doubts, you can just use some n-gram toolkit and compare it with the n-garm perplexity.

**Comment by ahmedalbahnasawy** [comment]
[pp.txt](https://github.com/espnet/espnet/files/4420438/pp.txt)
sounds good. The best perplexity was at epoch 48 out of 50. I 'm thinking to resume training up to 100 epochs. is it good ?

**Comment by sw005320** [comment]
more matched-domain data is always helpful.
I think you may not have to train more epochs. It would be already converged.

**Comment by ahmedalbahnasawy** [comment]
currently I use the acoustic text to train the LM. If i want to re train my LM by another data from specific domain for example :health or science. Do you think i can customize my ASR model in this case ? Note : I have no acoustic data for those types of domain

**Comment by sw005320** [comment]
A straightforward way is to just train LM with the specific data. There are several studies to transfer the ASR model itself to this target domain only with text data, but it is in a research-level.

**Comment by ahmedalbahnasawy** [comment]
@sw005320 
Thank you so much
